File: python/src/dorado/cli/gyotaku.py
"""The gyotaku CLI: Skein-512 file hashing, like sha256sum but Skein. Streams in
constant memory. Named for the Japanese art of fish-printing: a file's fingerprint."""

from __future__ import annotations

import argparse
import sys

from .. import skein


def _out_len(bits: int) -> int:
    if bits <= 0 or bits % 8 != 0:
        raise ValueError("--bits must be a positive multiple of 8")
    return bits // 8


def _digest_file(path: str, out_len: int) -> str:
    with open(path, "rb") as f:
        return skein.hash_stream(out_len, f).hex()


def _format_line(tag: bool, digest: str, name: str) -> str:
    return f"SKEIN-512 ({name}) = {digest}" if tag else f"{digest}  {name}"


def _run_check(files: list[str]) -> int:
    checked = failed = 0
    for lst in files:
        with open(lst, "r") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                parsed = _parse_check_line(line)
                if parsed is None:
                    print(f"error: {lst}: malformed line: {line}", file=sys.stderr)
                    return 1
                expected, name = parsed
                checked += 1
                try:
                    got = _digest_file(name, len(expected) // 2)
                except OSError:
                    print(f"{name}: FAILED open or read")
                    failed += 1
                    continue
                if got.lower() == expected.lower():
                    print(f"{name}: OK")
                else:
                    print(f"{name}: FAILED")
                    failed += 1
    if checked == 0:
        print("error: no digests found to verify", file=sys.stderr)
        return 1
    if failed:
        print(f"error: {failed} of {checked} digests did NOT match", file=sys.stderr)
        return 1
    return 0


def _parse_check_line(line: str):
    if line.startswith("SKEIN-512 (") and ") = " in line:
        name = line[len("SKEIN-512 (") : line.rindex(") = ")]
        digest = line[line.rindex(") = ") + 4 :].strip()
        return digest, name
    parts = line.split(None, 1)
    if len(parts) == 2:
        return parts[0], parts[1].lstrip("*").strip()
    return None


def main(argv: list[str] | None = None) -> int:
    p = argparse.ArgumentParser(prog="gyotaku", description="Skein-512 file hashing (educational, unaudited).")
    p.add_argument("--bits", type=int, default=256, help="output length in bits (a multiple of 8)")
    p.add_argument("-c", "--check", action="store_true", help="verify checksum lists")
    p.add_argument("--tag", action="store_true", help='BSD-style output: "SKEIN-512 (file) = digest"')
    p.add_argument("files", nargs="*")
    a = p.parse_args(argv)

    try:
        if a.check:
            if a.tag:
                raise ValueError("--tag cannot be combined with --check")
            if not a.files:
                raise ValueError("--check needs at least one checksum file")
            return _run_check(a.files)

        out_len = _out_len(a.bits)
        if not a.files:
            digest = skein.hash_stream(out_len, sys.stdin.buffer).hex()
            print(_format_line(a.tag, digest, "-") if a.tag else digest)
        else:
            for path in a.files:
                print(_format_line(a.tag, _digest_file(path, out_len), path))
    except (OSError, ValueError) as e:
        print(f"error: {e}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: ts/src/cli/gyotaku.ts
// gyotaku: standalone Skein-512 hashing CLI (Node). Port of the Rust/Go tool.
import { readFileSync } from "node:fs";
import { parseArgs } from "node:util";
import process from "node:process";
import * as skein from "../skein";
import { bytesToHex } from "../bytes";

function readStdin(): Uint8Array {
  return new Uint8Array(readFileSync(0));
}

function digest(data: Uint8Array, outLen: number): string {
  return bytesToHex(skein.hash(outLen, data));
}

function tagLine(d: string, name: string): string {
  return `SKEIN-512 (${name}) = ${d}`;
}

function isEvenHex(s: string): boolean {
  return s.length > 0 && s.length % 2 === 0 && /^[0-9a-fA-F]+$/.test(s);
}

function parseCheckLine(line: string): [string, string] | null {
  const bsd = line.match(/^.+ \((.+)\) = ([0-9a-fA-F]+)$/);
  if (bsd && isEvenHex(bsd[2])) return [bsd[2], bsd[1]];
  const sp = line.indexOf(" ");
  if (sp < 0) return null;
  const d = line.slice(0, sp);
  if (!isEvenHex(d)) return null;
  const name = line.slice(sp + 1).trimStart().replace(/^\*/, "");
  return [d, name];
}

function runCheck(files: string[]): void {
  let checked = 0,
    failed = 0;
  for (const list of files) {
    for (const raw of readFileSync(list, "utf8").split("\n")) {
      const line = raw.trim();
      if (line === "" || line.startsWith("#")) continue;
      const parsed = parseCheckLine(line);
      if (!parsed) throw new Error(`${list}: malformed line: ${line}`);
      const [expected, name] = parsed;
      checked++;
      try {
        const got = digest(new Uint8Array(readFileSync(name)), expected.length / 2);
        if (got.toLowerCase() === expected.toLowerCase()) console.log(`${name}: OK`);
        else {
          console.log(`${name}: FAILED`);
          failed++;
        }
      } catch {
        console.log(`${name}: FAILED open or read`);
        failed++;
      }
    }
  }
  if (checked === 0) throw new Error("no digests found to verify");
  if (failed > 0) throw new Error(`${failed} of ${checked} digests did NOT match`);
}

const USAGE = `usage: gyotaku [flags] [FILE...]

Hash files (or stdin when no FILE is given) with Skein-512, like sha256sum.

  --bits N      output length in bits, a multiple of 8 (default 256)
  --tag         BSD-style output: "SKEIN-512 (file) = digest"
  -c, --check   read digests from the FILEs and verify them
  --version     print version and exit
  -h, --help    print this help and exit
`;

function main(): void {
  const { values, positionals } = parseArgs({
    options: {
      bits: { type: "string", default: "256" },
      tag: { type: "boolean", default: false },
      c: { type: "boolean", short: "c", default: false },
      check: { type: "boolean", default: false },
      version: { type: "boolean", default: false },
      help: { type: "boolean", short: "h", default: false },
    },
    allowPositionals: true,
  });
  if (values.help) {
    process.stdout.write(USAGE);
    return;
  }
  if (values.version) {
    console.log("gyotaku 0.1.0");
    return;
  }
  if (values.c || values.check) {
    if (values.tag) throw new Error("--tag cannot be combined with -c");
    if (positionals.length === 0) throw new Error("-c needs at least one checksum file");
    runCheck(positionals);
    return;
  }
  const bits = parseInt(values.bits, 10);
  if (!Number.isInteger(bits) || bits <= 0 || bits % 8 !== 0) throw new Error("--bits must be a positive multiple of 8");
  const outLen = bits / 8;

  if (positionals.length === 0) {
    const d = digest(readStdin(), outLen);
    console.log(values.tag ? tagLine(d, "-") : d);
    return;
  }
  for (const path of positionals) {
    const d = digest(new Uint8Array(readFileSync(path)), outLen);
    console.log(values.tag ? tagLine(d, path) : `${d}  ${path}`);
  }
}

try {
  main();
} catch (e) {
  console.error("error:", e instanceof Error ? e.message : e);
  process.exit(1);
}
